### DISCUSSION TITLE
How to write a sync command?

### DISCUSSION THREAD

**Discussion by lexicalunit** [Rapptz/discord.py#8170] [Q&A] [answered]
How are we supposed to write a sync command to sync application commands? First you have to clear the commands, then sync them. However after you clear commands, you can't restore the tree to then be able to actually sync your commands.

```python
bot: MyBot # subclass of discord.ext.commands.Bot, 
           # uses cogs to create a bunch of commands
           # these cogs are then loaded in its setup_hook()

# settings.DEBUG_GUILD is `None` (global) or some int (a debug guild id)
guild_id: Optional[int] = settings.DEBUG_GUILD

if guild_id:
    print("syncing commands to debug guild:", guild_id)
else:
    print("syncing global commands")

guild = discord.Object(id=guild_id) if guild_id else None

# ok, let's first clear commands!
bot.tree.clear_commands(guild=guild)
await bot.tree.sync(guild=guild)

# cool, they've been cleared, now let's sync them!
# oh wait, we can't, they've been cleared... there's no way to restore bot.tree :(
await bot.tree.sync(guild=guild)
```

**Comment by LeoCx1000** [comment]
You DONT clear. Syncing is an upsert operation. It'll update discord's list of command to be synced with yours. Meaning it will delete, create and/or update commands.

**Reply by lexicalunit** [reply]
If I don't clear I get duplicate commands. Multiple `/whatever` commands only one of which actually work. I have to clear.

**Reply by scarletcafe** [reply]
I don't know exactly what's going on here, but this is likely a bug with the client and NOT actually duplicate commands getting uploaded. Syncing normally has worked for everyone else, and we've had a lot of people do it.

**Reply by lexicalunit** [reply]
According to this library's author I am supposed to be clearing commands: https://github.com/Rapptz/discord.py/discussions/7884 -- So I don't think I'm doing anything out of the ordinary here. I'm glad whatever you and "everyone else" are doing is working for you, but it's clearly not working nor is it intended to work the way you think it should work. Otherwise the `clear_commands()` method wouldn't even exist in the first place.

**Reply by Rapptz** [reply]
I did not suggest clearing commands every time you synced. You might have misread me. You had guild scoped and global scoped commands set which is why there was a duplicate. I suggested clearing one of them (i.e. the guild scoped ones) so the duplicates do not show up. Other than that, global commands cannot be duplicated so the clearing in the typical case is unnecessary.

**Reply by scarletcafe** [reply]
Given that I've been a long standing contributor to the library and helped people in the Discord for a number of years, I can assure you that "the way [I] think it should work" *is* the way it is intended to work. I am assuring you here that what you are doing here is unnecessary, even if the cause wasn't what I guessed it to be from the limited information available in this thread.

Personally, I think the non-functional overlapping of identical guild and global commands *should* be considered a Discord bug, but it's not really within the library's agency to address and well, Discord has never really listened to what we think anyway.

Please reconsider looking over the information again and revising your implementation. It's clear that you don't think you should need to take excess effort to do this - and I agree. If nothing else, I'll take this into consideration in terms of trying to ensure it gets covered in the [docs project](https://github.com/Rapptz/discord.py/issues/7654).

**Reply by lexicalunit** [reply]
Ah, I didn't realize one of the duplicates was the global and the other was for the debug guild. I'm migrating my codebase from the (not great, IMHO) discord-interactions library to discord.py 2.0. Back when I thought my bot would suddenly stop working unless I switched over to slash commands, I evaluated all my options and thought that discord-interactions seemed the best of the handful of bad options I was given.

I got slash commands working in time for Discord's original deadline, but then work on discord.py started again, and then the deadlines got pushed back by Discord. So now I have the time to convert over to discord.py 2.0 and rip out this other library. Which is great. Absolutely, it is.

But I never had the dupe issue when using discord-interactions so... I dunno what they are doing different, but it always _just worked_ with that library. I figured I must be doing something wrong. That I was using discord.py 2.0 incorrectly somehow.

So this is all I know: The solution I present as the selected answer in this discussion seems to work flawlessly. And if there's another solution that would also work flawlessly, using discord.py 2.0, I still haven't found it.

**Comment by lexicalunit** [comment, marked as answer]
Hopefully this is helpful to someone else, but this is basically what I did to solve this:

```python
import asyncio
import logging
from importlib import import_module
from inspect import isclass
from pathlib import Path
from pkgutil import iter_modules
from typing import TYPE_CHECKING

from discord.ext import commands
from discord.ext.commands import Bot

from mybot.cogs import FooCog
from mybot.cogs import BarCog
from mybot.cogs import BazCog
# ...

if TYPE_CHECKING:
    from mybot.client import MyBot

logger = logging.getLogger(__name__)

__all__ = [
    "FooCog",
    "BarCog",
    "BazCog",
    #...
]

async def load_all_cogs(bot: Bot) -> Bot:
    # iterate through the modules in the current package
    package_dir = Path(__file__).resolve().parent
    for info in iter_modules([str(package_dir)]):

        # import the module and iterate through its attributes
        module = import_module(f"{__name__}.{info.name}")
        for attribute_name in dir(module):  # pragma: no cover
            attribute = getattr(module, attribute_name)

            # Only load cogs in this module if they're exported
            if (
                isclass(attribute)
                and issubclass(attribute, commands.Cog)
                and attribute.__name__ in __all__
            ):
                if module.__name__ in bot.extensions:
                    logger.info("reloading extension %s...", module.__name__)
                    await bot.reload_extension(module.__name__)
                else:
                    logger.info("loading extension %s...", module.__name__)
                    await bot.load_extension(module.__name__)
                break
    return bot

async def load_extensions(bot: MyBot, do_sync: bool = False) -> None:
    guild = bot.settings.GUILD_OBJECT

    if do_sync:
        if guild:
            logger.info("syncing commands to debug guild: %s", guild.id)
        else:
            logger.info("syncing global commands")

        logger.info("clearing commands...")
        bot.tree.clear_commands(guild=guild)
        await bot.tree.sync(guild=guild)

        logger.info("waiting to avoid rate limit...")
        await asyncio.sleep(1)

    logger.info("loading cogs...")
    await load_all_cogs(bot)
    commands = [c.name for c in bot.tree.get_commands(guild=guild)]
    logger.info("registered commands: %s", ", ".join(commands))

    if do_sync:
        logger.info("syncing commands...")
        await bot.tree.sync(guild=guild)

# Then, inside MyBot#setup_hook:
await load_extensions(self)

# And inside the `!sync` command implementation in SyncCog:
await load_extensions(bot, do_sync=True)
```

I was hoping there was an easier way of doing this tho!
